python/lsst/ci/tsort.py

```diff
@@ -1,59 +1,103 @@
-#!/usr/bin/env python
-# -- mjuric: Taken from https://pypi.python.org/pypi/tsort and slightly modified
-# -- mjuric: LGPL licenced
-# -*- coding: utf-8 -*-
-
+from typing import Iterator, Iterable, List, Mapping, Set, Tuple
 
 class GraphError(Exception):
     pass
 
 
-def tsort(edges):
-    """topologically sort vertices in edges.
-    edges: list of pairs of vertices. Edges must form a DAG.
-           If the graph has a cycle, then GraphError is raised.
-    returns: topologically sorted list of vertices.
-    see http://en.wikipedia.org/wiki/Topological_sorting
+def to_dep_graph(edges: Iterable[Tuple[str, str]]) -> Mapping[str, Set[str]]:
+    """Takes an iterable collection of (object, dependency) pairs and returns
+    as a graph set"""
+    graph = {}
+    for node, dep in edges:
+        graph.setdefault(node, set()).add(dep)
+        # Ensure all deps are added to graph	
+        graph.setdefault(dep, set())
+    return graph
+
+
+def toposort(edges: Iterable[Tuple[str, str]]) -> Iterator[List]:
+    """Takes an iterable collection of (object, dependency) pairs
+    and returns an iterator of ordered dependency lists.
+    The items in each list can be processed in any order.
+    """
+    return toposort_mapping(to_dep_graph(edges))
+
+
+def toposort_dfs(edges: Iterable[Tuple[str, str]]) -> List[str]:
```

Comment:
Please clearly explain the difference between this and `toposort` above. Why would I chose one over the other? Why do both exist? The acronym "dfs" means nothing to me. I see it described as "depth-first sort" in a doc string below and that information needs to be in this doc string as well.

Reply:
This DFS variants are not currently used, but I'd like to keep them in the code for now. I've expanded much more on the difference between the two variants.